### PR TITLE
feat(dialog): add focus management

### PR DESCRIPTION
**Pull request by jelbourn**
R: @kara or @hansl 

This is the final PR necessary before publishing dialog.

(though there's still plenty to do after that)

### CODE REVIEW

File: src/lib/dialog/dialog-container.ts [hunk 1]
```diff
@@ -23,23 +27,52 @@ import {MdDialogContentAlreadyAttachedError} from './dialog-errors';
...
   attachTemplatePortal(portal: TemplatePortal): Map<string, any> {
     throw Error('Not yet implemented');
   }
+
+  ngOnDestroy() {
+    // When the dialog is destroyed, return focus to the element that originally had it before
+    // the dialog was opened. Wait for the DOM to finish settling before changing the focus so
+    // that it doesn't end up back on the <body>.
+    this._ngZone.onMicrotaskEmpty.first().subscribe(() => {
+      (this._elementFocusedBeforeDialogWasOpened as HTMLElement).focus();
```
Comment (kara) [inline src/lib/dialog/dialog-container.ts]: Do you prefer `as HTMLElement` to `(<HTMLElement>this.whatever).focus()` ?

File: src/lib/dialog/dialog-container.ts [hunk 2]
```diff
@@ -23,23 +27,52 @@ import {MdDialogContentAlreadyAttachedError} from './dialog-errors';
...
   attachComponentPortal<T>(portal: ComponentPortal<T>): ComponentRef<T> {
     if (this._portalHost.hasAttached()) {
       throw new MdDialogContentAlreadyAttachedError();
     }
 
-    return this._portalHost.attachComponentPortal(portal);
+    let attachResult = this._portalHost.attachComponentPortal(portal);
+
+    // If were to attempt to focus immediately, then the content of the dialog would not yet be
+    // ready in stances where change detection has to run first. To deal with this, we simply
```
Comment (kara) [inline src/lib/dialog/dialog-container.ts]: `in stances` -> ` in instances`

File: src/lib/dialog/dialog.spec.ts [hunk 3]
```diff
@@ -100,6 +107,55 @@ describe('MdDialog', () => {
...
+    it('should focus the first tabbable element of the dialog on open', fakeAsync(() => {
+      let config = new MdDialogConfig();
+      config.viewContainerRef = testViewContainerRef;
+
+      dialog.open(PizzaMsg, config);
+      viewContainerFixture.detectChanges();
+      flushMicrotasks();
+
+      expect(document.activeElement.tagName)
+          .toBe('INPUT', 'Expected the first tabbale element (input) in the dialog to be focused.');
```
Comment (kara) [inline src/lib/dialog/dialog.spec.ts]: `tabbale` -> `tabbable`

File: src/lib/dialog/dialog-container.ts (hunk 1 above)
Comment (jelbourn) [inline src/lib/dialog/dialog-container.ts]: I didn't know the `as` syntax existed until recently; I do this it is a bit cleaner and more in keeping with TypeScript's normally putting the type at the end. We can chat about it when we're all back in the office.

File: src/lib/dialog/dialog.spec.ts (hunk 3 above)
Comment (jelbourn) [inline src/lib/dialog/dialog.spec.ts]: Done

File: src/lib/dialog/dialog-container.ts (hunk 2 above)
Comment (jelbourn) [inline src/lib/dialog/dialog-container.ts]: Done

File: src/lib/dialog/dialog-container.ts (hunk 1 above)
Comment (kara) [inline src/lib/dialog/dialog-container.ts]: Don't feel strongly about it; was just curious

**Pull request merged**